src/essence/Tools/Isochrone/IsochroneTool_Algorithm.js

```diff
@@ -0,0 +1,172 @@
+            if(childIndex !== length - 1 && this.compare(childIndex, childIndex + 1))
+                childIndex++;
+            if(this.compare(index, childIndex)) {
+                this.swap(index, childIndex);
+                swapped = true;
+                index = childIndex;
+                childIndex = h_getChildIndex(index);
+            }
+        }
+    }
+
+    insert(px) {
+        let index = this.contents.length;
+        this.contents.push(px);
+        setPx(this.indexArr, px, index);
+        this.bubbleUp(index);
+    }
+
+    remove() {
+        const result = this.contents[0];
+        setPx(this.indexArr, result, -1);
+        const lastPx = this.contents.pop();
+        if(this.contents.length > 0) {
+            this.contents[0] = lastPx;
+            setPx(this.indexArr, lastPx, 0);
+            this.bubbleDown();
+        }
+        return result;
+    }
+}
+
+function pxToLatLng(px, tileBounds, zoom) {
+    const point = tileBounds.min
+        .multiplyBy(256)
+        .add([px[_X] + 0.5, px[_Y] + 0.5]);
+    return Map_.map.unproject(point, zoom);
+}
+
+//Clockwise from the right
+function moveToBacklink(move) {
+    return [
+        [ 0,  4,  0,  6,  0],
+        [ 2,  3,  5,  7,  8],
+        [ 0,  1,  0,  9,  0],
+        [16, 15, 13, 11, 10],
+        [ 0, 14,  0, 12,  0]
+    ][move[_Y] + 2][move[_X] + 2];
+}
+
+const queenMoves = [
+    [-1, -1], [-1, 0], [-1, 1],
+    [0, -1],          [0, 1],
+    [1, -1], [1, 0], [1, 1]
+];
+
+const knightMoves = [
+    ...queenMoves,
+    [-2, -1], [-2, 1],
+    [-1, -2], [-1, 2],
+    [1, -2], [1, 2],
+    [2, -1], [2, 1]
+];
+
+export default function generateIsochrone(
+    startPx, //L.Point
+    bounds, //L.Bounds
+    costFunction,
+    zoom,
+    maxCost = Infinity,
+    knightsMove = true
+) {
+    const moves = knightsMove ? knightMoves : queenMoves;
+
+    const width = (bounds.max.x - bounds.min.x) * 256;
+    const height = (bounds.max.y - bounds.min.y) * 256;
+
+    //Set up dijkstra's
+    const makeDataArray = (type, fillWith) => {
+        let arr = [];
+        for(let y = 0; y < height; y++) {
+            arr.push(new type(width).fill(fillWith));
+        }
+        return arr;
+    }
+    let costArr = makeDataArray(Float32Array, Infinity);
+    let linkArr = makeDataArray(Uint8Array, 0);
+    let indexArr = makeDataArray(Int32Array, -2);
+        //-2: px is unvisited; -1: px has been removed
+    
+    let heap = new PxHeap(costArr, indexArr);
+
+    costArr[startPx.y][startPx.x] = 0;
+    heap.insert([startPx.y, startPx.x]);
+
+    //Do dijkstra's to it
+    while(heap.contents.length > 0) {
+        const cPx = heap.remove(); //current pixel
+        const cCost = getPx(costArr, cPx);
+        const cLatLng = pxToLatLng(cPx, bounds, zoom);
+        for(const m of moves) {
+            const tPx = [cPx[_Y] + m[_Y], cPx[_X] + m[_X]]; //target pixel
+            if(tPx[_Y] < 0 || tPx[_X] < 0 || tPx[_Y] >= height || tPx[_X] >= width)
```

Comment:
for these kind of complex looking conditionals, it's a good idea to leave a comment explaining what it means in english